### ISSUE TITLE
Proxying SSE in IIS buffers

### ISSUE THREAD

**Issue by 1337bacon** [microsoft/reverse-proxy#533]
### What should we add or change to make your life better?
Websocket protocol support, add SignalR support

### Why is this important to you?
So the proxy has support for the actual websocket protocol not just the inital http handshake.

First of, I'm a bit of a noob when it comes to websockets. So please bare with me.
So I'd need a long living websocket connection.(for events send from the server)

So from what I could gather from the source code this proxy has websockets supported (or at least the handshake).
The initial handshake goes over HTTP. When the client gets an OK from the server a websocket connection is established. 
So correct me if I'm wrong, but the connection is established between the client and server directly without the proxy.
Because the proxy only handles http requests and not websocket protocol.

Looks like SignalR has some features that could come in handy.
Like managing multiple connections.

TLDR: Is there a way to proxy the actual websocket protocol? Is this even  good practice?

**Comment by Tratcher** [comment]
> So correct me if I'm wrong, but the connection is established between the client and server directly without the proxy.
> Because the proxy only handles http requests and not websocket protocol.

Not quite. The WebSocket response is a 101 which tells all parties to re-use the existing connection through the proxy as an opaque, bidirectional stream. WebSockets happens over that stream and the proxy doesn't need to no any additional details.

WebSockets and SignalR should work today.

**Comment by davidfowl** [comment]
SignalR requires stickyness, we should test that out with YARP (I know we support that)

**Comment by Tratcher** [comment]
Ah, I didn't distinguish between the SignalR handshake and the WebSocket handshake. Yes, if you have more than one destination instance in your cluster then you need to enable [session affinity](https://microsoft.github.io/reverse-proxy/articles/session-affinity.html) to make sure both requests end up going to the same instance, but both should still be happening through the proxy.

**Comment by 1337bacon** [comment]
Thank you @Tratcher and  @davidfowl for the quick reply.
Seems obvious now...
I've checked the source again, I now see  I missed the key step where streams are copied from client to destination streams and vice versa. This makes the whole thing much clearer now.

**Comment by 1337bacon** [comment]
One more thing.
We are in the process of migrating from our current proxy to YARP. 
We think WebSockets are the way to go, but we are also using SSE(server side events).
As far as I understand YARP does not support it. 
Could you please let me know if that is the case?

**Comment by davidfowl** [comment]
> As far as I understand YARP does not support it.

Why wouldn't YARP support SSE?

**Comment by 1337bacon** [comment]
Sorry for the late response.
At our current proxy implementation(not yarp) we have a separate class handling incoming SSE. Basically it copies the incoming stream to an output stream. Similar to what yarp is doing with websockets, but in one direction only.  So I assumed there must be something similar here. 
Also SSE does not work over YARP with our server/client SSE. BUT it does work with a simplified example of client/server which I used for testing.
It appears there must be some weirdness on our side...

Thanks for now.

**Comment by 1337bacon** [comment]
So I've finally made it work. 
I'll leave this here for anyone else having similar issues.

The first issue was that I messed up something while testing(mentioned above).
I changed some code then the examples stopped working.
The client only got messages from the server after the connection was closed.
So I assumed that has something to do with response buffering.
Now, before I pass the request to the proxy, I call `context.Features.Get<IHttpResponseBodyFeature>().DisableBuffering();`.
This seems to the trick.
Proxy is running on IIS (not IIS Express).

**Comment by Tratcher** [comment]
> Now, before I pass the request to the proxy, I call `context.Features.Get<IHttpResponseBodyFeature>().DisableBuffering();`.

@davidfowl this is because of IIS's dynamic compression messing up SSE? YARP should probably always disable this, or give you the option to do so.

**Comment by BrennanConroy** [comment]
> IIS's dynamic compression messing up SSE

Yeah I think so.

Maybe only disable for "text/event-stream"? Or are there more cases and it should be applied more generally

**Comment by davidfowl** [comment]
Shouldn't we always disable it?

**Comment by Tratcher** [comment]
You'd loose some functionality like dynamic compression. It seems like you'd want to be able to opt into our out of this setting.

**Comment by karelz** [comment]
Triage: Either workaround in YARP to disable buffering, or general configuration/ability to disable buffering per kind of stream / response.

**Comment by kikaragyozov** [comment]
@Tratcher I'm sorry for the necro, but I figured this was the best place to ask this as I was tackling issues with SSE on a live IIS server. Everything worked as expected in IIS Express/Kestrel/.NET Aspire, but the moment it got hosted in production, SSE stopped working.

After investigating, I thought the Dynamic Response Caching module was at fault here.

Even though I've configured the following in my `web.config`, the issue persisted:

```
<httpCompression>
  <dynamicTypes>
    <add mimeType="text/event-stream" enabled="false" />
  </dynamicTypes>
</httpCompression>
```

The only solution was to disable it via `HttpContext!.Features.Get<IHttpResponseBodyFeature>()!.DisableBuffering()`.

Could you please shine some light as to why the `web.config` approach didn't work, like, at all?


**EDIT:** After further investigations, it seems the dynamic compression module for IIS was **not even installed** on the system, so I'm not sure what's causing it to buffer responses. As previously mentioned - if the project is hosted on IIS Express/Kestrel or as part of orchestration support with .NET Aspire, there is no caching involved. If there is no dynamic compression involved in IIS, what else is there that could cause response buffering? 

I'm ghastly confused.

![image](https://github.com/microsoft/reverse-proxy/assets/6695731/a8cf4024-be01-4722-bb15-97ad91e18f11)

**Comment by Tratcher** [comment]
Comments on closed issues are not tracked, please open a new issue with the details for your scenario.

**Comment by kikaragyozov** [comment]
> Comments on closed issues are not tracked, please open a new issue with the details for your scenario.

Thank you, I've gone ahead and opened an issue here: https://github.com/dotnet/aspnetcore/issues/55654
